src/charm.py

```diff
@@ -14,18 +14,19 @@
 class FastAPIDemoCharm(ops.CharmBase):
     """Charm the service."""
 
-    def __init__(self, framework):
+    def __init__(self, framework: ops.Framework) -> None:
         super().__init__(framework)
         self.pebble_service_name = "fastapi-service"
-        framework.observe(self.on.demo_server_pebble_ready, self._on_demo_server_pebble_ready)
+        framework.observe(self.on["demo-server"].pebble_ready, self._on_demo_server_pebble_ready)
```

Comment:
> Note also the change to on["demo-server"].pebble_ready which provides better typing and imo should be the recommended way to access bound events for observe.

`ObjectEvents.__getitem__` returns a `PrefixedEvents` object and the `__getattr__` on that returns a `BoundEvent`. `ObjectEvents.__getattr__` returns `Any`, but if the framework is being used correctly, the only type of attributes that can be dynamically set is `EventSource`, where `__getattr__` returns `BoundEvent`.

If `ObjectEvents.__getattr__` had a `EventSource` return type, wouldn't the typing for both be equally good (which, given how dynamic everything is, really is actually not very good in either case)?

If, by providing `"demo-server"`, ops could somehow determine that this was a container and so the events would be only the workload ones (and similar for relations, storage, etc) then I think this would have more value.

I really dislike that we have two ways to do this and I'd like Charm-Tech to be consistent in using and recommending one. I lean towards `<container name>_pebble_ready` over `["<container name>"].pebble_ready` but I'm open to being convinced.

I think this would probably be a good team discussion, or we could potentially have it here or in Ops Library Development.

Reply:
> If `ObjectEvents.__getattr__` had a `EventSource` return type, wouldn't the typing for both be equally good

I think you're right that we could change the return type of that `__getattr__` to get typing parity between the two ways of getting events for `observe`, and perhaps we should do that in the near future.

Would it be a breaking change for anyone? Probably not, since `observe` should be the only place this is used, right? And its typing is of course compatible with this already.
```python
def observe(self, bound_event: BoundEvent, observer: Callable[[Any], None]):
    ...
```

---

Her's my pitch for the recommended method to be using the name as an index.

In terms of typing, `on["demo-server"]` is currently a bit of a halfway solution, because it still doesn't get you the actual event type. However, it's still preferable to the completely dynamic attribute lookup, since it takes us in the direction of only using static attribute names like `pebble_ready`, and clearly separates dynamic string shenanigans from regular attributes. 

Since `__getattr__` is only called if the attribute doesn't exist, `PrefixedEvents` could be updated to define a real attribute for all the known events, with annotations for their return types, which would provide exact event types for anyone using the `on["demo-server"]` form. This wouldn't save people getting the name wrong, or using a non-container name for a container event -- these would just blow up at runtime the same way getting them wrong currently would. But it would save them from typos in the event half of the name, or using an event that doesn't exist, in addition to exposing more precise type information when writing their `observe` calls.

Using dynamic names as string keys rather than integrating them into attribute names with hyphens substituted out also makes renaming the server container (etc) easier, especially if you're storing the names in variables (e.g. `on[SERVER_CONTAINER]`).

I think that having these dynamic attributes in the first place mostly make sense if we're thinking of `ops` as not necessarily needing to know that much about changes `juju` makes (e.g. adding new event types), but I don't think that's the direction `ops` has ended up in, and it would probably be a good idea to generally steer away from their use where possible.

---

>I think this would probably be a good team discussion

Agreed, let's bring it up at today's sync.

Reply:
As per discussion in today's sync, I'll revert to `self.on.demo_server_pebble_ready` until we have a more compelling case to recommend otherwise